From 95536a77d1581a10e69a8e263de7711df3ef77f3 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:33:30 -0400
Subject: [iommu] vt-d: release invalidation queue when destroying IOMMU unit

Message-id: <20140905013330.30867.6802.stgit@gir.stowe>
Patchwork-id: 91288
O-Subject: [RHEL7.1 PATCH 24/28] iommu/vt-d: release invalidation queue when destroying IOMMU unit
Bugzilla: 1129808
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Dutile <ddutile@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129808
Upstream Status: a84da70b7ba0c5236fccf25115acefc235ed65f9

commit a84da70b7ba0c5236fccf25115acefc235ed65f9
Author: Jiang Liu <jiang.liu@linux.intel.com>
Date:   Mon Jan 6 14:18:23 2014 +0800

    iommu/vt-d: release invalidation queue when destroying IOMMU unit

    Release associated invalidation queue when destroying IOMMU unit
    to avoid memory leak.

    Signed-off-by: Jiang Liu <jiang.liu@linux.intel.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/dmar.c b/drivers/iommu/dmar.c
index f564932..67473a8 100644
--- a/drivers/iommu/dmar.c
+++ b/drivers/iommu/dmar.c
@@ -738,6 +738,12 @@ static void free_iommu(struct intel_iommu *iommu)
   destroy_irq(iommu->irq);
  }
 
+ if (iommu->qi) {
+  free_page((unsigned long)iommu->qi->desc);
+  kfree(iommu->qi->desc_status);
+  kfree(iommu->qi);
+ }
+
  if (iommu->reg)
   unmap_iommu(iommu);
 
-- 
1.7.1